verifier/src/qe_report_body.rs

```diff
@@ -0,0 +1,548 @@
+        let attributes = self.attributes.verify(&qe_report_body.0);
+        let isv_svn = self.isv_svn.verify(&qe_report_body.0);
+        let status = mr_signer.is_success()
+            & isv_prod_id.is_success()
+            & miscellaneous_select.is_success()
+            & attributes.is_success()
+            & isv_svn.is_success();
+        VerificationOutput::new(
+            QeReportBodyValue {
+                mr_signer,
+                isv_prod_id,
+                miscellaneous_select,
+                attributes,
+                isv_svn,
+            },
+            status,
+        )
+    }
+}
+
+impl VerificationMessage<QeReportBodyValue> for QeReportBodyVerifier {
+    fn fmt_padded(
+        &self,
+        f: &mut Formatter<'_>,
+        pad: usize,
+        output: &VerificationOutput<QeReportBodyValue>,
+    ) -> core::fmt::Result {
+        let status = choice_to_status_message(output.is_success());
+
+        write!(
+            f,
+            "{:pad$}{status} QE Report Body all of the following must be true:",
+            ""
+        )?;
+        let pad = pad + MESSAGE_INDENT;
+        writeln!(f)?;
+        self.mr_signer.fmt_padded(f, pad, &output.value.mr_signer)?;
+        writeln!(f)?;
+        self.isv_prod_id
+            .fmt_padded(f, pad, &output.value.isv_prod_id)?;
+        writeln!(f)?;
+        self.miscellaneous_select
+            .fmt_padded(f, pad, &output.value.miscellaneous_select)?;
+        writeln!(f)?;
+        self.attributes
+            .fmt_padded(f, pad, &output.value.attributes)?;
+        writeln!(f)?;
+        self.isv_svn.fmt_padded(f, pad, &output.value.isv_svn)
+    }
+}
+
+/// Verifier for ensuring a QE(quoting enclave) ISV SVN falls within the
+/// provided [`TcbLevel`]s.
+///
+/// Preforms step 4 of
+/// <https://api.portal.trustedservices.intel.com/documentation#pcs-enclave-identity-v4>
+///
+/// > 4. Determine a TCB status of the Enclave:
+/// >       a. Retrieve a collection of TCB Levels (sorted by ISVSVNs) from
+/// >          tcbLevels field in Enclave Identity structure.
+/// >       b. Go over the list of TCB Levels (descending order) and find the
+/// >          one that has ISVSVN that is lower or equal to the ISVSVN value
+/// >          from SGX Enclave Report.
+/// >       c. If a TCB level is found, read its status from tcbStatus field,
+/// >          otherwise your TCB Level is not supported.
+#[derive(Debug, Clone)]
+struct QeIsvSvnVerifier {
+    tcb_levels: Vec<TcbLevel>,
+}
+
+impl QeIsvSvnVerifier {
+    fn new<'a, I>(tcb_levels: I) -> Self
+    where
+        I: IntoIterator<Item = &'a TcbLevel>,
+    {
+        Self {
+            tcb_levels: tcb_levels.into_iter().cloned().collect(),
+        }
+    }
+}
+
+impl<E: Accessor<IsvSvn>> Verifier<E> for QeIsvSvnVerifier {
+    type Value = (IsvSvn, TcbLevel);
+    fn verify(&self, evidence: &E) -> VerificationOutput<Self::Value> {
```

Comment:
Does this assume the tcb_levels list is correctly sorted?

Reply:
yes it does, the step does say `(sorted by ISVSVNs)` so I think it's should be safe and potentially more robust to sort here again.  I'll work on that

Reply:
Add logic to ensure the TCB levels are sorted.